### PR TITLE
Added CI based on GitHub Actions.

### PR DESCRIPTION
**Pull request by thresheek**
Pushes to master and stable branches will result in buildbot-like checks on multiple OSes and architectures.

Pull requests will be checked on a public Ubuntu 22.04 GitHub runner.

### CODE REVIEW

File: .github/workflows/check-pr.yml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+on:
+  pull_request:
+
+jobs:
+  build-and-test:
+    runs-on: [ ubuntu-22.04 ]
+    steps:
+      - name: checkout v4
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
```
Comment (defanator) [inline .github/workflows/check-pr.yml]: any reason to stick to particular git sha? why not just `actions/checkout@v4`?

File: .github/workflows/check-pr.yml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+      - name: Check out nginx tests
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
```
Comment (defanator) [inline .github/workflows/check-pr.yml]: ditto

File: .github/workflows/check-pr.yml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            --with-stream_geoip_module \
+            --with-cc-opt="$(dpkg-buildflags --get CFLAGS)" \
+            --with-ld-opt="$(dpkg-buildflags --get LDFLAGS)" \
+          || cat objs/autoconf.err
+          make -j$(nproc) -k || make
+
+      - name: Test
+        working-directory: nginx-tests
+        run: |
+          prove -v -j$(nproc) --state=save ../t . || prove -v --state=failed
```
Comment (defanator) [inline .github/workflows/check-pr.yml]: any worth in saving that state as build artifact?

File: .github/workflows/buildbot.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+on:
+  push:
+    branches:
+      - master
+      - 'stable-1.*'
+
+jobs:
+  buildbot:
+    uses: nginx/ci-self-hosted/.github/workflows/nginx-buildbot.yml@main
```
Comment (pluknet) [inline .github/workflows/buildbot.yml:11]: You include buildbot.yml jobs logic from the "ci-self-hosted" separate repo.
Why not do the same for check-pr.yml to reduce unrelated changes in the "nginx" repo.

**Review by pluknet — COMMENTED** [review]
I don't know the github actions logic, the rest looks good.

File: .github/workflows/check-pr.yml (hunk 1 above)
Comment (thresheek) [reply .github/workflows/check-pr.yml]: It's a best practice to use specific sha ids for actions so we don't come up with unexpected breaking changes if they're introduced by actions authors.

That is to say, we need to bump this to the current version and test.

File: .github/workflows/check-pr.yml (hunk 3 above)
Comment (thresheek) [reply .github/workflows/check-pr.yml]: Probably not.  We don't really do that on buildbot now and rely on reproducing it manually.

File: .github/workflows/buildbot.yml (hunk 4 above)
Comment (thresheek) [reply .github/workflows/buildbot.yml:11]: That's a good idea, I'll explore it in the next version of PR some time next week.

File: .github/workflows/check-pr.yml (hunk 1 above)
Comment (defanator) [reply .github/workflows/check-pr.yml]: uhm, any references for this? I'm seeing this advice for 3rd party actions only:

https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

Comment (defanator) [reply .github/workflows/check-pr.yml]: (we use standard actions a lot, and github tends to warn you if you're using something old that is about to turn obsolete - right at workflow job results page)

Comment (thresheek) [reply .github/workflows/check-pr.yml]: I don't see why we should apply a different strategy for second-party (GH) actions as opposed to third-party (anyone else).  It's still not something we directly control.

File: .github/workflows/check-pr.yml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            zlib1g-dev \
+            ffmpeg \
+            libcache-memcached-perl \
+            libcryptx-perl \
+            libio-socket-ssl-perl \
+            libtest-harness-perl \
+            libprotocol-websocket-perl \
+            libscgi-perl \
+            uwsgi \
+            uwsgi-plugin-python3
```
Comment (defanator) [inline .github/workflows/check-pr.yml]: Also, it's worth to consider [caching](https://github.com/actions/cache) all these to speed up regular workflows.

File: .github/workflows/check-pr.yml (hunk 1 above)
Comment (defanator) [reply .github/workflows/check-pr.yml]: okay, fair enough (but still weird)

**Review by oxpa — APPROVED** [review]

**Review by pluknet — CHANGES_REQUESTED** [review]
Set as a reminder: pending move GitHub Actions logic to ci-self-hosted.

File: .github/workflows/check-pr.yml (hunk 5 above)
Comment (thresheek) [reply .github/workflows/check-pr.yml]: Do you suggest we should cache the downloaded .deb files prior to the install or something else?

File: .github/workflows/buildbot.yml (hunk 4 above)
Comment (thresheek) [reply .github/workflows/buildbot.yml:11]: Done in https://github.com/nginx/nginx/commit/1b022486987ece6e2127d6e3c1f839b4ae32073f

**Comment by thresheek** [comment]
I've pushed a new version of a patch to move the jobs logic to a separate repo fully.

File: .github/workflows/check-pr.yml (hunk 5 above)
Comment (defanator) [reply .github/workflows/check-pr.yml]: You can cache the results of the above installations and reuse it later I guess. Saves some time on (re)downloading + installing, esp. on workflows that run relatively often (I would expect good amount of activity here in future).

Comment (defanator) [reply .github/workflows/check-pr.yml]: Again, this is not a blocker, just suggestion. We can track runners' usage to improve things as we go.

**Comment by defanator** [comment]
Uhm,

> Pull requests will be checked on a public Ubuntu 22.04 GitHub runner.

Wouldn't it be better to keep that workflow open to the public, actually?

**Review by pluknet — APPROVED** [review]

**Comment by thresheek** [comment]
> Uhm,
> 
> > Pull requests will be checked on a public Ubuntu 22.04 GitHub runner.
> 
> Wouldn't it be better to keep that workflow open to the public, actually?

Those workflows are public.

File: .github/workflows/check-pr.yml (hunk 5 above)
Comment (thresheek) [reply .github/workflows/check-pr.yml]: Sure, I just don't really understand how we can do it for debian packages.  We not only need to cache the debs, and cache where it's installed filesystem-wide, we also need to re-run the post-scripts for those packages.  That seems like too much effort to save 20-30 seconds of install time.

**Pull request merged**